libkmod/libkmod-elf.c

```diff
@@ -126,8 +126,12 @@ static inline uint64_t elf_get_uint(const struct kmod_elf *elf, uint64_t offset,
 		for (i = 0; i < size; i++)
 			ret = (ret << 8) | p[i];
 	} else {
+#if __BYTE_ORDER == __LITTLE_ENDIAN
+		memcpy(&ret, p, size);
+#else
```

Comment:
Added, although I would argue with size here for now. We have tons of unneeded checks which weigh heavily on runtime performance.

Reply:
The issue is that the real potential can only be seen if these checks are gone and this code is optimized by compiler. But having both in one PR makes it really hard to justify if everything's correct. So ... yeah, I have linked a branch in PR description where a good work in progress can be seen, but it's really hard to verify its correctness (even there, still too many checks included).

Thanks for your feedback anyway!

Reply:
I'm ok with optimizing some places for "same-endianess", but  it needs to be done generically instead of adding ifdefs in the middle of the functions.  So here you could either extract a copy function that takes the elf->msb into consideration or... just always use memcpy and if endianness does not match, you swap the bytes - the source location may not be aligned, but the dest is.

Reply:
How should I know if endianess does not match if it's not an information the compiler delivers? Doing some kind of runtime check with a union to see which byte is set feels very wrong.

Reply:
Ah, got it. `<endian.h>` is key here. This removes preprocessor instructions.

Reply:
And adjusted for better C readability. Compilers still optimize it with default options as good as before.